waiter/waiter.go

```diff
@@ -0,0 +1,41 @@
+package waiter
+
+import (
+	"fmt"
+	"math"
+	"time"
+)
+
+// ComputeDelay will compute delay between waiter attempts using the min delay, max delay, remaining time and
+// current attempt count.  The attempt count is the request call count.
+// The zeroth attempt takes no delay.
+func ComputeDelay(minDelay, maxDelay, remainingTime time.Duration, attempt int64) (time.Duration, error) {
+	// validation
+	if minDelay > maxDelay {
+		return 0, fmt.Errorf("maximum delay must be greater than minimum delay")
+	}
+
+	// zeroth attempt, no delay
+	if attempt <= 0 {
+		return 0, nil
+	}
+
+	// remainingTime is zero or less, no delay
+	if remainingTime <= 0 {
+		return 0, nil
+	}
+
+	// [0.0, 1.0) * 2 ^ attempt-1
+	ri := 1 << uint64(attempt-1)
+
+	delay := time.Duration(math.Min(
+		maxDelay.Seconds(),
+		minDelay.Seconds()*float64(ri)),
+	) * time.Second
```

Comment:
This is susceptible to integer overflow if number of attempts is large enough.

To compute the max attempt ceiling for max delay this should use Log2, and then check the number of attempts against the max attempt ceiling.

```go
// Get attempt ceiling to prevent integer overflow.
attemptCeiling := math.Log2(maxDelay.Seconds())
if attempt > int(attemptCeiling) {
     attempt = attemptCeiling
}

// Compute exponential delay based on attempt.
ri := 1 << uint64(attempt-1)
delaySeconds: = float(ri)

// remaining work
if remainingTime-delay <= minDelay {
	delay = remainingTime - minDelay
}

return delay, nil
```

Reply:
Algorithm was wrong for attempt ceiling its should be, 

It should be `minDelay * 2^x = maxDelay`.  Which to solve for x need to do

```
attemptCeiling = (math.Log(float64(maxDelay / minDelay)) / math.Log(2)) + 1
```

https://play.golang.org/p/wgLPZ1rF5Hr

Reply:
Also since this would be using Log, something should ensure that `minDelay` and `maxDelay` are greater than zero.